8325304: Several classes in java.util.jar and java.util.zip don't specify the behaviour for null arguments

jaikiran opened the pull request:
Can I please get a review of this doc-only change which updates the javadoc of several classes in `java.util.jar` and `java.util.zip` to specify their behaviour when `null` arguments are passed to the constructor or methods of those classes?

For these updated classes, I have individually checked that they indeed throw a `NullPointerException` when `null` is passed to their constructor or methods. The couple of places where `null` is accepted have been updated to mention that `null` is allowed.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8325305](https://bugs.openjdk.org/browse/JDK-8325305) to be approved

### Issues
 * [JDK-8325304](https://bugs.openjdk.org/browse/JDK-8325304): Several classes in java.util.jar and java.util.zip don't specify the behaviour for null arguments (**Bug** - P4)
 * [JDK-8325305](https://bugs.openjdk.org/browse/JDK-8325305): Several classes in java.util.jar and java.util.zip don't specify the behaviour for null arguments (**CSR**)


### Reviewers
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17728/head:pull/17728` \
`$ git checkout pull/17728`

Update a local copy of the PR: \
`$ git checkout pull/17728` \
`$ git pull https://git.openjdk.org/jdk.git pull/17728/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17728`

View PR using the GUI difftool: \
`$ git pr show -t 17728`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17728.diff">https://git.openjdk.org/jdk/pull/17728.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17728#issuecomment-1929183035)

jaikiran commented:
/csr needed

jaikiran commented:
A CSR has been created for this change https://bugs.openjdk.org/browse/JDK-8325305

LanceAndersen reviewed (commented):
Hi Jai,

Thank you for tackling this.  An initial comment on a quick read

I noticed you use "can cause" here and later use "will cause".  I think we should try to be consistent.

Perhaps we should use "will result in a..."

jaikiran commented:
Hello Lance,

> I noticed you use "can cause" here and later use "will cause". I think we should try to be consistent.
> Perhaps we should use "will result in a..."

My use of "can cause" was indeed inconsistent. I had used it in a couple of places. I have now updated the PR to replace with "will cause".

I decided to use this "will cause" for consistency, since the text:

> "Unless otherwise noted, passing a {@code null} argument to a constructor or method in this class will cause a {@link NullPointerException} to be thrown." 

has been already in use in various other classes (for example `JarFile`'s javadoc) in the JDK even before the proposed changes in this PR.

AlanBateman commented:
The changes means that NPE is specified in both the class description and method descriptions in some cases, e.g. JarInputStream. I assume you meant to remove the NPE from methods where you've added a statement in the class description.

Minor nit is some of the usages of <p> in the change are inconsistent with the usages close by, you probably want to keep them consistent if you can.

jaikiran commented:
Hello Alan,

> The changes means that NPE is specified in both the class description and method descriptions in some cases, e.g. JarInputStream. I assume you meant to remove the NPE from methods where you've added a statement in the class description.

I wasn't sure if I should remove them from the pre-existing method javadocs, so I had left them without changes. I have now updated the PR to remove them from the method level javadoc where we have now added the class level specification.

> 
> Minor nit is some of the usages of `<p>` in the change are inconsistent with the usages close by, you probably want to keep them consistent if you can.

Updated the PR to address this as well. I have updated it to match whatever the existing style (if any) was in that particular file.

AlanBateman commented on src/java.base/share/classes/java/util/zip/DeflaterInputStream.java:
```diff
@@ -75,6 +75,9 @@ private void ensureOpen() throws IOException {
     public DeflaterInputStream(InputStream in) {
+        // "in" being null isn't allowed. we use a null check for "in"
+        // merely to avoid an unnecessary instance creation of the Deflater
+        // for such erroneous cases.
```
I see this same comment has been added in 3 places but it's not easy to read and I don't think is needed.

jaikiran replied:
I've now updated the PR to remove this code comment.

AlanBateman commented on src/java.base/share/classes/java/util/zip/CheckedOutputStream.java:
```diff
@@ -39,10 +39,13 @@
+     * Creates an output stream with the specified Checksum. A null
+     * value to either {@code out} or {@code cksum} will cause
+     * a {@link NullPointerException} to be thrown from the
+     * {@code write} methods of this {@code CheckedOutputStream}.
```
What is the reason for specifying the NPE in the method description rather than a throws?

jaikiran replied:
These 2 classes, the `CheckedInputStream` and the `CheckedOutputStream` are slightly different from the rest of the classes in this changeset. This javadoc here is for the constructor of the `CheckedInputStream`. The implementation of the constructor just blindly assigns the passed argument to an internal member field and doesn't do any null check on the passed arguments. So it isn't throwing any `NullPointerException` here. So adding a `@throws` here would be incorrect. In theory, we could just change the implementation of this constructor to throw a `NullPointerException` if either of these arguments were null, but that might mean a potential breakage of some weird usage of the CheckedInputStream. So I decided to add the `NullPointerException` detail to the constructor description.

Do you think we should instead do something like this for this class and the `CheckedOutputStream` class:

```diff
diff --git a/src/java.base/share/classes/java/util/zip/CheckedInputStream.java b/src/java.base/share/classes/java/util/zip/CheckedInputStream.java
index 10f72b416d1..76cba26986e 100644
--- a/src/java.base/share/classes/java/util/zip/CheckedInputStream.java
+++ b/src/java.base/share/classes/java/util/zip/CheckedInputStream.java
@@ -41,10 +41,7 @@ public class CheckedInputStream extends FilterInputStream {
     private final Checksum cksum;
 
     /**
-     * Creates an input stream using the specified Checksum. A null
-     * value to either {@code in} or {@code cksum} will cause
-     * a {@link NullPointerException} to be thrown from the
-     * {@code read} methods of this {@code CheckedInputStream}.
+     * Creates an input stream using the specified Checksum.
      * @param in the input stream
      * @param cksum the Checksum
      */
@@ -57,6 +54,8 @@ public CheckedInputStream(InputStream in, Checksum cksum) {
      * Reads a byte. Will block if no input is available.
      * @return the byte read, or -1 if the end of the stream is reached.
      * @throws    IOException if an I/O error has occurred
+     * @throws    NullPointerException if this {@code CheckedInputStream} was
+     *            constructed with a {@code null} value for {@code in} or {@code cksum}
      */
     public int read() throws IOException {
         int b = in.read();
@@ -75,7 +74,9 @@ public int read() throws IOException {
      * @param len the maximum number of bytes read
      * @return    the actual number of bytes read, or -1 if the end
      *            of the stream is reached.
-     * @throws     NullPointerException If {@code buf} is {@code null}.
+     * @throws     NullPointerException If {@code buf} is {@code null} or
+     *             if this {@code CheckedInputStream} was
+     *             constructed with a {@code null} value for {@code in} or {@code cksum}.
      * @throws     IndexOutOfBoundsException If {@code off} is negative,
      * {@code len} is negative, or {@code len} is greater than
      * {@code buf.length - off}


```

jaikiran replied:
Given that subclasses could set these fields lazily (however remote the case might be), do you think we should then not specify the `NullPointerException` for the read methods on these 2 classes. In which case I can exclude these 2 classes from this PR.

AlanBateman replied:
It would be simpler to drop CheckedInputStream and CheckedOutputStream and deal with them separately. It's unfortunate that this current effort has run into this but that comes with touching some of the JDK 1.0/1.1 era classes.

jaikiran replied:
Done. I've updated the PR to undo the changes that were proposed for `CheckedOutputStream` and `CheckedInputStream`. I've also updated the CSR with the latest specification diff from this PR.

AlanBateman reviewed (commented):
This mostly looks okay, just a few comments.

jaikiran commented:
Thank you Alan and Lance for the reviews. tier1, tier2 and tier3 testing went fine for this change.

jaikiran commented:
/integrate
